File: 4038-count-integers-appearing-in-a-single-block/4038-count-integers-appearing-in-a-single-block.py
class Solution:
    def countSpecialIntegers(self, nums: list[int]) -> int:
        if not nums:
            return 0

        total_count = Counter(nums)
        block_len = {}

        i = 0
        n = len(nums)
        while i < n:
            curr_num = nums[i]
            curr_len = 0
            while i < n and nums[i] == curr_num:
                curr_len += 1
                i += 1
            block_len[curr_num] = curr_len
        special_count = 0
        for num in total_count:
            if total_count[num] == block_len.get(num , 0):
                special_count += 1
        return special_count

File: 4038-count-integers-appearing-in-a-single-block/4038-count-integers-appearing-in-a-single-block.cpp
class Solution {
public:
    int countSpecialIntegers(vector<int>& nums) {
        if(nums.empty()){
            return 0;
        }

        unordered_map<int , int> total_count;
        unordered_set<int> d_nums;
        unordered_map<int , int> block_len;

        for(int num : nums){
            total_count[num]++;
            d_nums.insert(num);
        }

        int i = 0;
        int n = nums.size();
        while(i < n){
            int curr_num = nums[i];
            int curr_len = 0;

            while(i < n && nums[i] == curr_num){
                curr_len++;
                i++;
            }
            block_len[curr_num] = curr_len;
        }

        int special_count = 0;
        for(int num : d_nums){
            if(total_count[num] == block_len[num]){
                special_count++;
            }
        }

        return special_count;
    }
};